test/plugins/TPSetBufferCreator.cpp

```diff
@@ -0,0 +1,336 @@
+      } else {
+        frag_out->set_error_bit(dataformats::FragmentErrorBits::kDataNotFound, true);
+      }
+
+      send_out_fragment(std::move(frag_out));
+      sentCount++;
+      it++;
+    }
+  }
+
+  m_tps_buffer->clear_buffer(); // emptying buffer
+
+  TLOG() << get_name() << ": Exiting do_stop() method : sent " << sentCount << " incomplete fragments";
+}
+
+void
+TPSetBufferCreator::do_scrap(const nlohmann::json& /*args*/)
+{
+  TLOG() << get_name() << ": Entering do_scrap() method";
+  m_tps_buffer.reset(nullptr); // calls dtor
+  TLOG() << get_name() << ": Exiting do_scrap() method";
+}
+
+std::unique_ptr<dataformats::Fragment>
+TPSetBufferCreator::convert_to_fragment(TPSetBuffer::data_request_output /* ds_output */,
+                                        dfmessages::DataRequest input_data_request)
+{
+  auto ret = std::make_unique<dataformats::Fragment>(std::vector<std::pair<void*, size_t>>());
+  auto& frag = *ret.get();
+
+  dataformats::GeoID geoid(dataformats::GeoID::SystemType::kDataSelection, m_conf.region, m_conf.element);
+  dataformats::FragmentHeader frag_h;
+  frag_h.trigger_number = input_data_request.trigger_number;
+  frag_h.trigger_timestamp = input_data_request.trigger_timestamp;
+  frag_h.window_begin = input_data_request.window_begin;
+  frag_h.window_end = input_data_request.window_end;
+  frag_h.run_number = input_data_request.run_number;
+  frag_h.element_id = geoid;
+  frag_h.fragment_type = (dataformats::fragment_type_t)dataformats::FragmentType::kTriggerPrimitives;
+  frag_h.sequence_number = input_data_request.sequence_number;
+
+  frag.set_header_fields(frag_h);
+
+  /*  >>> need to fill the fragment with
+      the contents of the TPSet vector here
+   */
+
+  return ret;
+}
+
+void
+TPSetBufferCreator::send_out_fragment(std::unique_ptr<dataformats::Fragment> frag_out,
+                                      size_t& sentCount,
+                                      std::atomic<bool>& running_flag)
+{
+  std::string thisQueueName = m_output_queue_frag->get_name();
+  bool successfullyWasSent = false;
+  // do...while so that we always try at least once to send
+  // everything we generate, even if running_flag is changed
+  // to false between the top of the main loop and here
+  do {
+    TLOG() << get_name() << ": Pushing the requested TPSet onto queue " << thisQueueName;
+    try {
+      m_output_queue_frag->push(std::move(frag_out), m_queueTimeout);
+      successfullyWasSent = true;
+      ++sentCount;
+    } catch (const dunedaq::appfwk::QueueTimeoutExpired& excpt) {
+      std::ostringstream oss_warn;
+      oss_warn << "push to output queue \"" << thisQueueName << "\"";
+      ers::warning(dunedaq::appfwk::QueueTimeoutExpired(ERS_HERE, get_name(), oss_warn.str(), m_queueTimeout.count()));
+    }
+  } while (!successfullyWasSent && running_flag.load());
+}
+
+void
+TPSetBufferCreator::send_out_fragment(std::unique_ptr<dataformats::Fragment> frag_out)
```

Comment:
True, but I'm not sure how to set a default value to a std::atomic<bool>& argument in a function. Any suggestion?

Reply:
I think you can do it with an extra layer of indirection. Put the body of the `do...while` loop in its own function, which returns true if the send succeeded, and false otherwise, then call that function in a loop from your two `send_out_fragment` functions